### ISSUE TITLE
Decreasing number of running processes in non-task mode

### ISSUE THREAD

**Issue by gaow** [vatlab/SoS#971]
It will take some work to prepare a MWE, but just in case this is something obvious from the code I'm going to report my problem here for now. Running large number of small benchmark-style jobs with `concurrent = True` in non-task mode, I noticed that the workflow starts with multiple threads, then the number of concurrent threads drop as the workflow executes further, until at last it dropped to computing using only one thread towards the end. I recall we have implemented some file-based system to control for the number of concurrent jobs. Are there any switches or debug information somewhere that I can use to check if the CPU slots have been wrongfully reserved when there are in fact available slots?

**Comment by gaow** [comment]
Okay my simulations have been running on single thread for the past few hours now. This is really weird but I can now come up with this MWE here:

[slow_test.zip](https://github.com/vatlab/SoS/files/2007338/slow_test.zip)


It is a very simple pipeline. To reproduce, first run `top` in a separate screen to monitor it, then run

```
sos run parallel_stuck.sos -j 20
```

You'll see that for workflow `step01` completed rather quickly, but  `step02` got stuck and seems to be working with single thread -- you do not see a lot on `top`. If you let it run for a while with `step02`, then kill it and run again, you'll get text like this on the screen:

```
INFO: Executing step02_0: 
INFO: input:   [file_target('slow_test/step01/step01_1.rds'), ...] (100 items)
INFO: step02_0 (index=0) is ignored due to saved signature
INFO: step02_0 (index=1) is ignored due to saved signature
INFO: step02_0 (index=2) is ignored due to saved signature
INFO: step02_0 (index=3) is ignored due to saved signature
^CProcess ForkPoolWorker-1:20:
```

and each of these lines take a second to pop up, as if done on a single core.

Do you get the same behavior on your end?

**Comment by BoPeng** [comment]
The zip file is corrupted. You can try to gmail it to me.

**Comment by gaow** [comment]
Sorry, this one here should be good:

[15-May-2018.zip](https://github.com/vatlab/SoS/files/2007408/15-May-2018.zip)

It's a very small test but has a lot small tasks. My actual simulation has even more tasks. 

BTW, restarting the computer getting rid of cache, other running `R` processes etc did not help.

**Comment by BoPeng** [comment]
First, checking signature is slow, .. second, the jobs are submitted one by one and because the jobs are completed quickly, the queue is never filled and appear to be executed one by one.

**Comment by gaow** [comment]
Okay, but I did not understand why the first batch seems to end rather quickly. Rerunning them also will lead to signature check but they do seem to complete in parallel and within seconds. Only at the 2nd step the slow signature check seems to become problem. I find it hard to believe that the slowness is from signature check. Isn't that at least done in parallel?

I'm trying to resume my actual simulation job that got disrupted due to my reboot. I do not think it has even finished signature check now that 30min has passed, but again I find it unusual to see only a couple of the "ignored due to saved signature" stuff pop up every second.

**Comment by BoPeng** [comment]
Still checking.

**Comment by BoPeng** [comment]
I suspect it was due to some large variable in sig file.

**Comment by gaow** [comment]
Well, indeed every time I do load that `IO_DB` file as global variable. I was always wondering if there is a way to prevent that, in this and other applications ... Otherwise the variables involved should be small in size for this toy data. In my simulation case all I'm passing along are just parameters, too. Here is a list of files in my simulation job folder:

```
-rw-r--r-- 1 gaow gaow    0 May 15 19:30 9e5404641ff1e8c0.sig_
-rw-r--r-- 1 gaow gaow    0 May 15 19:30 716d89fb1acbe32a.sig_
-rw-r--r-- 1 gaow gaow    0 May 15 19:30 e3466f192df53812.sig_
-rw-r--r-- 1 gaow gaow    0 May 15 19:31 fdc1c2b25afd0813.sig_
-rw-r--r-- 1 gaow gaow    0 May 15 19:32 7d017f5c4894553b.sig_
-rw-r--r-- 1 gaow gaow  621 May 15 19:34 266da5a66844fa9c.sig
-rw-r--r-- 1 gaow gaow    0 May 15 19:35 6971a17f51b183bb.sig_
-rw-r--r-- 1 gaow gaow  12K May 15 19:35 6971a17f51b183bb.sig
-rw-r--r-- 1 gaow gaow    0 May 15 19:35 2dc40264861c1eb9.sig_
-rw-r--r-- 1 gaow gaow  21K May 15 19:35 2dc40264861c1eb9.sig
-rw-r--r-- 1 gaow gaow    0 May 15 19:35 f1111679da577e41.sig_
-rw-r--r-- 1 gaow gaow 504K May 15 19:36 f1111679da577e41.sig
-rw-r--r-- 1 gaow gaow    0 May 15 19:36 0f76dfdc48aaa1c0.sig_
-rw-r--r-- 1 gaow gaow 2.2M May 15 21:59 0f76dfdc48aaa1c0.sig
-rw-r--r-- 1 gaow gaow 2.8K May 15 22:30 9e5404641ff1e8c0.sig
-rw-r--r-- 1 gaow gaow 1.9K May 15 22:30 fdb79b216b81900f.sig
-rw-r--r-- 1 gaow gaow  29K May 15 22:30 716d89fb1acbe32a.sig
-rw-r--r-- 1 gaow gaow  50K May 15 22:30 e3466f192df53812.sig
-rw-r--r-- 1 gaow gaow 1.2M May 15 22:31 fdc1c2b25afd0813.sig
-rw-r--r-- 1 gaow gaow 2.5M May 15 22:55 7d017f5c4894553b.sig
```

The last one, `7d017f5c4894553b.sig`, is growing. I believe currently the workflow is still checking signatures.

**Comment by BoPeng** [comment]
The problem, at least one of the problems is 

```
output: step02_output_files, group_by = 1
```
SoS would have to check a large number of files (~4000 of them) for each input group....

**Comment by BoPeng** [comment]
This patch should help a bit...

**Comment by gaow** [comment]
Great! Just tried out the patch. There is an improvement for that MWE, but for my simulation case it is still quite slow -- at least I can still follow the signature check speed with my eyeballs. The step of the application I'm running involves 15K output files in groups of 2. 

> SoS would have to check a large number of files (~4000 of them) for each input group....

Check signature of ~4000 files for each input group? That should explain the slowness.

**Comment by BoPeng** [comment]
No, for this particular problem, the function merely tries to check if the file exists and if `glob.glob` should be called to expand `*` etc. The problem is that checking the existence of files in your case means `~4000` file system access for each input group. The patch takes the advantage of the fact that in `output` (not `input`), checking the existence of files is not needed so it only checks the existence of `*` and `?` in filenames.

**Comment by gaow** [comment]
I see. But the fact that `step02` has significantly bigger overhead than `step01` makes me suspect that something with `input` is dragging it slow ...

**Comment by BoPeng** [comment]
Actually, even without the patch,

```
output: step02_output_files[_index] 
```
would be much faster if it is what you need.

**Comment by BoPeng** [comment]
Do not know about your `input`, but at least for your MWE signature check is almost fast enough and I can see that all four slots are used by step 2.

**Comment by gaow** [comment]
Okay, let me try the `_index` trick first. But this will restrict myself to output `group_by=1` right? And I suppose I can do something like `output_files[2*_index:(2*_index+2)]` instead of `group_by=2` if there are multiple outputs per group?

I brought up `input` because am I right to worry that there will be the same burden to check existence of all input files for every input group?

**Comment by BoPeng** [comment]
`input` is different because it will be processed only once although the existence of input files have to be checked. `output` is executed for each input group and that is why you see the slowness.

The `_index` trick is not recommended if the patch more or less works. Because the `output` will be evaluated each time, I have always recommended to generate `_output` from `_input`, not as a subset of `step_output`.

**Comment by gaow** [comment]
Okay, then `input` should not be an issue ... this is good to know. 

As for `output`, the `_index` trick is a lot faster and finally my simulation continues to run! This is great. However I'm afraid that the patch did not help much for my case, at least after applying it my benchmark still stuck at signature checks. 

> Because the output will be evaluated each time, I have always recommended to generate _output from _input, not as a subset of step_output.

How can this be done for the type of workflow structure I use in generating benchmarks, as the MWE shows, since the `input` name is different from `output`, both pre-specified? In that case, the best I can do is to use `output_files[_index]`?

**Comment by BoPeng** [comment]
Perhaps I can optimize `group_by=1` and translate that into `[_index]`, but the basic idea is that passing too many files to `output` is not a good idea.
